### PR TITLE
feat: sync url with filters

### LINKED ISSUE #145: Sync chain and time filters with URL
Sync explorer filters with URL
    - Current behavior - search query is included
    - Desired behavior - origin, destination, time filters are included in the URL

### PR DESCRIPTION
**Pull request by Xaroz**
fixes #145 
- Chain picker and time picker will now sync with the URL. 
- Navigation to this URL will make it so that these filters will remain and be sent to the message query
- For the chains, the keys are `origin` and `destination` and their value will be a `domainId`. Inputting an invalid `domainId` will prompt the `SearchChainError` component to show
- For the time, the keys are `startTime` and `endTime` and their value is a number. The initial load of the values are validated by the function `tryToDecimalNumber()`
- Refactor functions to allow multiple query params.

### CODE REVIEW

File: src/components/search/SearchStates.tsx [hunk 1]
```diff
@@ -118,3 +118,14 @@ export function SearchUnknownError({ show }: { show: boolean }) {
     />
   );
 }
+
+export function SearchChainError({ show }: { show: boolean }) {
+  return (
+    <SearchError
+      show={show}
+      imgSrc={ErrorIcon}
+      text="Sorry, the origin or destination chain is invalid. Please try choosing another chain or cleaning your filters."
```
Comment (jmrossy) [inline src/components/search/SearchStates.tsx:127]:
```suggestion
      text="Sorry, either the origin or the destination chain is invalid. Please choose a different chain or clear your search filters."
```

File: src/features/messages/MessageSearch.tsx [hunk 2]
```diff
@@ -96,7 +140,16 @@ export function MessageSearch() {
...
           />
         </div>
-        <Fade show={!isAnyError && isValidInput && isAnyMessageFound && !!multiProvider}>
+        <Fade
+          show={
+            !isAnyError &&
+            isValidInput &&
+            isValidOrigin &&
+            isValidDestination &&
+            isAnyMessageFound &&
```
Comment (jmrossy) [inline src/features/messages/MessageSearch.tsx:149]: Long list, should we make a var above to combine these?

File: src/features/messages/queries/useMessageQuery.ts [hunk 3]
```diff
@@ -21,6 +22,11 @@ export function isValidSearchQuery(input: string) {
   return !!searchValueToPostgresBytea(input);
 }
 
+export function isValidDomainId(domainId: string | null, multiProvider: MultiProvider) {
+  if (!domainId) return false;
+  return !!multiProvider.tryGetDomainId(domainId);
```
Comment (jmrossy) [inline src/features/messages/queries/useMessageQuery.ts:27]:
```suggestion
  return multiProvider.hasChain(domainId);
```

File: src/utils/queryParams.ts [hunk 4]
```diff
@@ -15,28 +15,51 @@ export function getQueryParamString(query: ParsedUrlQuery, key: string, defaultV
...
     );
-    if (value) newQuery.set(key, value);
-    else newQuery.delete(key);
-    const path = `${pathname}?${newQuery.toString()}`;
-    router
-      .replace(path, undefined, { shallow: true })
-      .catch((e) => logger.error('Error shallow updating url', e));
+    Object.entries(params).forEach(([key, value]) => {
+      if (value) {
+        if (newQuery.get(key) !== value) {
```
Comment (jmrossy) [inline src/utils/queryParams.ts:43]: Awkward structure, combine these

File: src/utils/queryParams.ts [hunk 5]
```diff
@@ -15,28 +15,51 @@ export function getQueryParamString(query: ParsedUrlQuery, key: string, defaultV
...
+          newQuery.set(key, value);
+          hasChanged = true;
+        }
+      } else {
+        if (newQuery.has(key)) {
```
Comment (jmrossy) [inline src/utils/queryParams.ts:48]: Ditto

**Review by jmrossy — COMMENTED** [review]
Mostly looks good!

File: src/components/search/SearchStates.tsx (hunk 1 above)
Comment (Xaroz) [reply src/components/search/SearchStates.tsx]: done

File: src/features/messages/MessageSearch.tsx (hunk 2 above)
Comment (Xaroz) [reply src/features/messages/MessageSearch.tsx]: created a `showMessageTable` variable to combine these

File: src/features/messages/queries/useMessageQuery.ts (hunk 3 above)
Comment (Xaroz) [reply src/features/messages/queries/useMessageQuery.ts]: Didn't know there was a function for this, thanks!

File: src/utils/queryParams.ts (hunk 4 above)
Comment (Xaroz) [reply src/utils/queryParams.ts]: done

**Review by jmrossy — APPROVED** [review]

**Pull request merged**